### ISSUE TITLE
Using Bert embedding in dialog_qa

### ISSUE THREAD

**Issue by Haoyulance** [allenai/allennlp#2727]
```
{
    "dataset_reader": {
        "type": "quac",
        "lazy": true,
        "num_context_answers": 2,
        "token_indexers": {
            "bert": {
                "type": "bert-pretrained",
                "pretrained_model": "bert-large-cased"
            },
            "token_characters": {
                "type": "characters",
                "character_tokenizer": {
                    "byte_encoding": "utf-8",
                    "end_tokens": [
                        260
                    ],
                    "start_tokens": [
                        259
                    ]
                },
                "min_padding_length": 5
            }
        }
    },
    "iterator": {
        "type": "bucket",
        "batch_size": 10,
        "max_instances_in_memory": 1000,
        "sorting_keys": [
            [
                "question",
                "num_fields"
            ],
            [
                "passage",
                "num_tokens"
            ]
        ]
    },
    "model": {
        "type": "dialog_qa",
        "dropout": 0.2,
        "initializer": [],
        "marker_embedding_dim": 10,
        "num_context_answers": 2,
        "phrase_layer": {
            "type": "gru",
            "bidirectional": true,
            "hidden_size": 100,
            "input_size": 1144,  // elmo (1024) + cnn (100) + num_context_answers (2) * marker_embedding_dim (10)
            "num_layers": 1
        },
        "residual_encoder": {
            "type": "gru",
            "bidirectional": true,
            "hidden_size": 100,
            "input_size": 200,
            "num_layers": 1
        },
        "span_end_encoder": {
            "type": "gru",
            "bidirectional": true,
            "hidden_size": 100,
            "input_size": 400,
            "num_layers": 1
        },
        "span_start_encoder": {
            "type": "gru",
            "bidirectional": true,
            "hidden_size": 100,
            "input_size": 200,
            "num_layers": 1
        },
        "text_field_embedder": {
            "bert": {
                "type": "bert-pretrained",
                "pretrained_model": "bert-large-cased"

            },
            "allow_unmatched_keys": true,
            "embedder_to_indexer_map": {
                "bert": ["bert", "bert-offsets"],
                "token_characters": ["token_characters"],
            },
            "token_characters": {
                "type": "character_encoding",
                "dropout": 0.2,
                "embedding": {
                    "embedding_dim": 20,
                    "num_embeddings": 262
                },
                "encoder": {
                    "type": "cnn",
                    "embedding_dim": 20,
                    "ngram_filter_sizes": [
                        5
                    ],
                    "num_filters": 100
                }
            }
        }
    },
    "train_data_path": "https://s3.amazonaws.com/my89public/quac/train_5000.json",
    "validation_data_path": "https://s3.amazonaws.com/my89public/quac/val.json",
    "trainer": {
        "cuda_device": 0,
        "learning_rate_scheduler": {
            "type": "reduce_on_plateau",
            "factor": 0.5,
            "mode": "max",
            "patience": 3
        },
        "num_epochs": 30,
        "optimizer": {
            "type": "sgd",
            "lr": 0.01,
            "momentum": 0.9
        },
        "patience": 10,
        "validation_metric": "+f1"
    },
    "validation_iterator": {
        "type": "bucket",
        "batch_size": 3,
        "max_instances_in_memory": 1000,
        "sorting_keys": [
            [
                "question",
                "num_fields"
            ],
            [
                "passage",
                "num_tokens"
            ]
        ]
    }
}

```
I wanted to substitute the **Elmo** with **Bert** to train **dialog_qa** model. I changed  "*token_indexers*" and "*text_field_embedder*" parts with corresponding codes of Bert. However, I encountered some dimension error: 

```
  Traceback (most recent call last):
  File "/home/syyaoxin321/.conda/envs/allennlp/bin/allennlp", line 10, in <module>
    sys.exit(run())
  File "/home/syyaoxin321/.conda/envs/allennlp/lib/python3.6/site-packages/allennlp/run.py", line 18, in run
    main(prog="allennlp")
  File "/home/syyaoxin321/.conda/envs/allennlp/lib/python3.6/site-packages/allennlp/commands/__init__.py", line 101, in main
    args.func(args)
  File "/home/syyaoxin321/.conda/envs/allennlp/lib/python3.6/site-packages/allennlp/commands/train.py", line 103, in train_model_from_args
    args.force)
  File "/home/syyaoxin321/.conda/envs/allennlp/lib/python3.6/site-packages/allennlp/commands/train.py", line 136, in train_model_from_file
    return train_model(params, serialization_dir, file_friendly_logging, recover, force)
  File "/home/syyaoxin321/.conda/envs/allennlp/lib/python3.6/site-packages/allennlp/commands/train.py", line 204, in train_model
    metrics = trainer.train()
  File "/home/syyaoxin321/.conda/envs/allennlp/lib/python3.6/site-packages/allennlp/training/trainer.py", line 480, in train
    train_metrics = self._train_epoch(epoch)
  File "/home/syyaoxin321/.conda/envs/allennlp/lib/python3.6/site-packages/allennlp/training/trainer.py", line 322, in _train_epoch
    loss = self.batch_loss(batch_group, for_training=True)
  File "/home/syyaoxin321/.conda/envs/allennlp/lib/python3.6/site-packages/allennlp/training/trainer.py", line 263, in batch_loss
    output_dict = self.model(**batch)
  File "/home/syyaoxin321/.conda/envs/allennlp/lib/python3.6/site-packages/torch/nn/modules/module.py", line 489, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/syyaoxin321/.conda/envs/allennlp/lib/python3.6/site-packages/allennlp/models/reading_comprehension/dialog_qa.py", line 197, in forward
    embedded_question = self._text_field_embedder(question, num_wrapping_dims=1)
  File "/home/syyaoxin321/.conda/envs/allennlp/lib/python3.6/site-packages/torch/nn/modules/module.py", line 489, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/syyaoxin321/.conda/envs/allennlp/lib/python3.6/site-packages/allennlp/modules/text_field_embedders/basic_text_field_embedder.py", line 125, in forward
    return torch.cat(embedded_representations, dim=-1)
RuntimeError: invalid argument 0: Sizes of tensors must match except in dimension 3. Got 11 and 14 in dimension 2 at /pytorch/aten/src/THC/generic/THCTensorMath.cu:83
```

I'm wondering do I need to adjust the "**input_size**" which is 1144 for Elmo? I still can't find the reason of causing this problem and need some help. Thanks!!!

**Comment by matt-gardner** [comment]
I'm not familiar enough with the BERT code to have a good idea of what's going on here.  It _looks_ like you're getting a different number of tokens for BERT than you have for the token characters.  Is this because BERT adds tokens that aren't getting added for the token characters representation?  I didn't see anything obvious in the code that made me think that was what was happening.  Can you print the tokens that are used by each indexer?  Something is off here, and I'm not sure what it is.

**Comment by matt-gardner** [comment]
In talking this over with the team, we think this would be fixed by #2537.

**Comment by Haoyulance** [comment]
Thank you so much for your prompt responses! I will try to fix it by reading that content, thanks again!
